### PR TITLE
fix: issue with JSON conversion of packagings data

### LINKED ISSUE #752: Product.dart does not convert packagings correctly to JSON
### Description

Conversion of `Product` objects to JSON fails to generate valid JSON for the `packagings` field when this list/field contains `ProductPackaging` objects. In this case, the output generated using the `Product.toJson` function is invalid JSON and `Product` objects cannot be restored from this JSON using the `Product.fromJson` function. 

### Expected behavior

The following test should pass (but currently it does not with version 2.6.0 of the SDK).

```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:openfoodfacts/openfoodfacts.dart';

void main() {
  test('ProductPackaging bug', () {
    final testProduct = Product();
    const productName = 'Orange';
    testProduct.productName = productName;
    testProduct.packagings = [ProductPackaging()..quantityPerUnit = '75cl'];
    final productJson = testProduct.toJson();
    try {
      final productRestored = Product.fromJson(productJson);
      expect(productRestored.productName, productName);
    } catch (e, s) {
      print(s);
      fail('Could not restore Product from JSON ${e.toString()}');
    }
  });
}

```

### Stacktraces
If applicable, provide error stacktraces.

```
#0      new ProductPackaging.fromJson (package:openfoodfacts/src/model/product_packaging.dart:53:34)
#1      _$ProductFromJson.<anonymous closure> (package:openfoodfacts/src/model/product.g.dart:119:41)
#2      MappedListIterable.elementAt (dart:_internal/iterable.dart:415:31)
#3      ListIterator.moveNext (dart:_internal/iterable.dart:344:26)
#4      new _GrowableList._ofEfficientLengthIterable (dart:core-patch/growable_array.dart:189:27)
#5      new _GrowableList.of (dart:core-patch/growable_array.dart:150:28)
#6      new List.of (dart:core-patch/array_patch.dart:47:28)
#7      ListIterable.toList (dart:_internal/iterable.dart:214:7)
#8      _$ProductFromJson (package:openfoodfacts/src/model/product.g.dart:120:12)
#9      new Product.fromJson (package:openfoodfacts/src/model/product.dart:535:28)
#10     main.<anonymous closure> (file:///home/peter/git/fit/fitness_tracker/test/product-bug_test.dart:12:39)
#11     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:215:19)
<asynchronous suspension>
#12     StackZoneSpecification._registerUnaryCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:124:15)
<asynchronous suspension>
```

### Package information
Open Food Facts package version: 2.6.0

### PR DESCRIPTION
**Pull request by peterwvj**
Fixes #752

After adding the proposed fix, I did run into some issues regenerating the *.g.dart files using build_runner but I managed to fix that by upgrading the SDK dependencies according to the instructions described [here](https://github.com/dart-lang/build/issues/3519#issuecomment-1587514460).

### CODE REVIEW

File: test/issue752_test.dart [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    const productName = 'Orange';
+    testProduct.productName = productName;
+    testProduct.packagings = [ProductPackaging()..quantityPerUnit = '75cl'];
+    final productJson = testProduct.toJson();
+    try {
+      final productRestored = Product.fromJson(productJson);
+      expect(productRestored.productName, productName);
+    } catch (e, s) {
+      print(s);
+      fail('Could not restore Product from JSON ${e.toString()}');
```
Comment (g123k) [inline test/issue752_test.dart:16]: FYI, just printing `$e` is enough. It's implicit to call `toString`

Comment (peterwvj) [reply test/issue752_test.dart]: Addressed in in commit 5f127c2. Thanks!

File: lib/src/model/product.dart [hunk 2]
```diff
@@ -319,7 +319,10 @@ class Product extends JsonObject {
   @JsonKey(name: 'packaging', includeIfNull: false)
   String? packaging;
 
-  @JsonKey(name: 'packagings', includeIfNull: false)
+  @JsonKey(
+      name: 'packagings',
+      includeIfNull: false,
+      toJson: JsonHelper.productPackagingsToJson)
```
[the remark this answers is not in the document]
Comment (peterwvj) [reply lib/src/model/product.dart]: Addressed in in commit 5f127c2. Thanks!

File: lib/src/model/product.g.dart
Comment (monsieurtanuki) [inline lib/src/model/product.g.dart:1]: Just put this back at the beginning:
```dart
// ignore_for_file: deprecated_member_use_from_same_package
// That's a bit ugly, but we need the previous line in order to pass the pub.dev
// tests, so put that line back after code generation.
```

File: test/issue752_test.dart [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import 'package:test/test.dart';
+
+void main() {
+  test('ProductPackaging bug', () {
+    final testProduct = Product();
+    const productName = 'Orange';
+    testProduct.productName = productName;
+    testProduct.packagings = [ProductPackaging()..quantityPerUnit = '75cl'];
+    final productJson = testProduct.toJson();
+    try {
```
Comment (monsieurtanuki) [inline test/issue752_test.dart:11]: Would be easier to let the test fail naturally with an exception, but why not.

File: test/issue752_test.dart [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      final productRestored = Product.fromJson(productJson);
+      expect(productRestored.productName, productName);
```
Comment (monsieurtanuki) [inline test/issue752_test.dart:13]: You should test `packagings` too.

**Review by monsieurtanuki — CHANGES_REQUESTED** [review]
Hi @peterwvj!
The code itself looks good, but you removed important `ignore` statements in generated files. Please put them back so that the code analyze works.
And consider my other comments ;)

**Comment by peterwvj** [comment]
@monsieurtanuki thank you for your great PR feedback. I've implemented all of your suggestions in commit df353be

**Review by monsieurtanuki — APPROVED** [review]
Thank you @peterwvj for this PR and the changes after my review!

**Pull request merged**